tasks/survey/minerprotocols/minerprotocols.go

```diff
@@ -0,0 +1,177 @@
+package minerprotocols
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/filecoin-project/go-address"
+	lapi "github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/gammazero/workerpool"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/multiformats/go-multiaddr"
+
+	"github.com/filecoin-project/lily/model"
+	observed "github.com/filecoin-project/lily/model/surveyed"
+)
+
+var (
+	resultBufferEnv  = "LILY_SURVEY_MINER_PROTOCOL_BUFFER"
+	resultBufferSize = 50
+
+	workerPoolSizeEnv = "LILY_SURVEY_MINER_PROTOCOL_WORKERS"
+	workerPoolSize    = 50
+
+	fetchTimeoutEnv = "LILY_SURVEY_MINER_PROTOCOL_TIMEOUT_SECONDS"
+	fetchTimeout    = 30
+)
+
+func init() {
+	if s := os.Getenv(resultBufferEnv); s != "" {
+		v, err := strconv.ParseInt(s, 10, 64)
+		if err == nil {
+			resultBufferSize = int(v)
+		}
+	}
+	if s := os.Getenv(workerPoolSizeEnv); s != "" {
+		v, err := strconv.ParseInt(s, 10, 64)
+		if err == nil {
+			workerPoolSize = int(v)
+		}
+	}
+	if s := os.Getenv(fetchTimeoutEnv); s != "" {
+		v, err := strconv.ParseInt(s, 10, 64)
+		if err == nil {
+			fetchTimeout = int(v)
+		}
+	}
+}
+
+type API interface {
+	Host() host.Host
+	StateListMiners(ctx context.Context, tsk types.TipSetKey) ([]address.Address, error)
+	StateMinerInfo(ctx context.Context, addr address.Address, tsk types.TipSetKey) (lapi.MinerInfo, error)
+}
+
+func NewTask(api API) *Task {
+	return &Task{api: api}
+}
+
+type Task struct {
+	api API
+}
+
+func (t *Task) Process(ctx context.Context) (model.Persistable, error) {
+	miners, err := t.api.StateListMiners(ctx, types.EmptyTSK)
+	if err != nil {
+		return nil, fmt.Errorf("listing miners: %w", err)
+	}
+
+	start := time.Now()
+	out := make(observed.MinerProtocolList, 0, len(miners))
+	results := make(chan *observed.MinerProtocol, resultBufferSize)
+	pool := workerpool.New(workerPoolSize)
+
+	for _, miner := range miners {
+		select {
+		case <-ctx.Done():
+			pool.Stop()
+			return nil, ctx.Err()
+		default:
+		}
+
+		miner := miner
+
+		// find the miner, if DNE abort as this indicates an error in the API as a miner was returned from StateListMiners that DNE in state tree.
+		// passing EmptyTSK causes API to use current chain head.
```

Comment:
Wouldn't it be possible for a miner to fall out of `StateListMiners` between tipsets? If we're always querying HEAD while walking through miners, this failure case may happen more often than we would expect. Would it be better to always query using the same tipset key?

And even if you disagree, I don't think it's appropriate to bail on the rest of the miner list, is it? Maybe just continue instead?

Reply:
Good catch, will fetch the current chain head at the start of the method and use it for all calls.